packages/mdc-select/index.ts

```
@@ -0,0 +1,596 @@
+    this.addMutationObserverForRequired_();
+  }
+
+  /**
+   * Initializes the select's event listeners and internal state based
+   * on the environment's state.
+   */
+  initialSyncWithDOM() {
+    this.handleChange_ = () => this.foundation_.handleChange(/* didChange */ true);
+    this.handleFocus_ = () => this.foundation_.handleFocus();
+    this.handleBlur_ = () => this.foundation_.handleBlur();
+    this.handleClick_ = (evt) => {
+      if (this.selectedText_) {
+        this.selectedText_.focus();
+      }
+      this.foundation_.handleClick(this.getNormalizedXCoordinate_(evt));
+    };
+    this.handleKeydown_ = (evt) => this.foundation_.handleKeydown(evt);
+    this.handleMenuSelected_ = (evtData) => this.selectedIndex = evtData.detail.index;
+    this.handleMenuOpened_ = () => {
+      // Menu should open to the last selected element.
+      if (this.selectedIndex >= 0) {
+        const selectedItemEl = this.menu_!.items[this.selectedIndex] as HTMLElement;
+        selectedItemEl.focus();
+      }
+    };
+    this.handleMenuClosed_ = () => {
+      // menuOpened_ is used to track the state of the menu opening or closing since the menu.open function
+      // will return false if the menu is still closing and this method listens to the closed event which
+      // occurs after the menu is already closed.
+      this.menuOpened_ = false;
+      this.selectedText_!.removeAttribute('aria-expanded');
+      if (document.activeElement !== this.selectedText_) {
+        this.foundation_.handleBlur();
+      }
+    };
+
+    // One of these elements must be non-null.
+    const element = this.nativeControl_ || this.selectedText_!;
+
+    element.addEventListener('change', this.handleChange_);
+    element.addEventListener('focus', this.handleFocus_);
+    element.addEventListener('blur', this.handleBlur_);
+
+    MOUSEDOWN_TOUCHSTART_EVENTS.forEach((evtType) => {
+      element.addEventListener(evtType, this.handleClick_ as EventListener);
```

since element can potentially be null, I think we need a guard here.

---

The component is only valid if nativeInput or selectedText is present, and the code above declares that selectedText will be defined if nativeInput isn't (there's also a comment above it).

---

`element` should never be null:

```ts
// One of these elements must be non-null.
const element = this.nativeControl_ || this.selectedText_!;
```

---

Should we then throw an error if both are missing? It seems like a helpful user error to know why the component wouldn't be working. I guess it is obvious if you don't have a select/list, just throwing out ideas.

---

Yeah, that sounds like a good idea. I'll do that and add a unit test.

---

Done